src/codec/dogstatsd.rs

```diff
@@ -0,0 +1,795 @@
+                let tags: Vec<&str> = section[1..].split(",").collect();
+                m.insert("tags".into(), Value::from(tags));
+            } else if section.starts_with('c') {
+                let container_id = &section[2..];
+                m.insert("container_id".into(), Value::from(container_id));
+            }
+        }
+    }
+
+    Ok(Value::from(m))
+}
+
+fn decode_service_check(data: &[u8]) -> Result<Value> {
+    let mut d = data.iter().enumerate();
+    let mut m = Object::with_capacity(8);
+    m.insert("dogstatsd_type".into(), Value::from("service_check"));
+    let start_index: usize;
+
+    // Skip the prefix and set the starting
+    loop {
+        match d.next() {
+            Some((idx, b'|')) => {
+                start_index = idx + 1;
+                break;
+            }
+            _ => (),
+        }
+    }
+
+    // Name
+    loop {
+        match d.next() {
+            Some((idx, b'|')) => {
+                let name = substr(data, start_index..idx)?;
+                m.insert("name".into(), Value::from(name));
+                break;
+            }
+            Some(_) => (),
+            None => return Err(invalid()),
+        }
+    }
+
+    // Status
+    match d.next() {
+        Some((idx, b'0' | b'1' | b'2' | b'3')) => {
+            let status_str = substr(data, idx..=idx)?;
+            let status: i32 = status_str.parse()?;
+            m.insert("status".into(), Value::from(status));
+        }
+        _ => return Err(invalid()),
+    }
+
+    // Optional Sections
+    match d.next() {
+        Some((idx, b'|')) => {
+            let sections: Vec<&str> = substr(data, idx + 1..)?.split("|").collect();
+            for section in sections.iter() {
+                if section.starts_with('d') {
+                    let timestamp: u32 = section[2..].parse()?;
+                    m.insert("timestamp".into(), Value::from(timestamp));
+                } else if section.starts_with('h') {
+                    let hostname = &section[2..];
+                    m.insert("hostname".into(), Value::from(hostname));
+                } else if section.starts_with('#') {
+                    let tags: Vec<&str> = section[1..].split(",").collect();
+                    m.insert("tags".into(), Value::from(tags));
+                } else if section.starts_with('m') {
+                    let message = &section[2..];
+                    m.insert("message".into(), Value::from(message));
+                } else if section.starts_with('c') {
+                    let container_id = &section[2..];
+                    m.insert("container_id".into(), Value::from(container_id));
+                }
+            }
+        }
+        Some(_) => return Err(invalid()),
+        None => (),
+    }
+
+    Ok(Value::from(m))
+}
+
+fn invalid() -> Error {
+    Error::from(ErrorKind::InvalidDogStatsD)
+}
+
+fn substr<I: SliceIndex<[u8], Output = [u8]>>(data: &[u8], r: I) -> Result<&str> {
```

Comment:
this function is copied from statsd and not changed at all. What about moving the statsd function to the tremor-common module and importing it there and here? The test below `test_subslice` could also be move to `tremor-common`

Reply:
Thank you for the walkthrough to make this work!